### DISCUSSION TITLE
How can I resize an image in-place?

### DISCUSSION THREAD

**Discussion by sampletext32** [SixLabors/ImageSharp#1946] [Ideas]
# Brief explanation

I have a custom allocated byte array, which is then wrapped into Image<Rgba32>.
When I try to resize this image, I get an exception. What I want to achieve is Image staying in it's original size, but the content is scaled.

Here is a quick resulting Image of what I want to achieve.
![image](https://user-images.githubusercontent.com/44116740/150163096-58dee3db-0175-4f39-8617-55a6a3217b62.png)

I am already aware of Cloning-Resizing-Drawing technique, but it's extremely memory consumptive.

**Comment by antonfirsov** [comment]
@sampletext32 unfortunately, our current Resize processor doesn't work inplace internally. It's always using separate source and destination buffers behind the scenes. When user is doing `Mutate`, the destination buffer is swapped out and becomes the new backing buffer of the image. This technique obviously doesn't work with wrapped-memory images -- that's why we throw the exception.

> Cloning-Resizing-Drawing

I would better recommend *Load -> Resize (Mutate) -> Copy to destination buffer* for your case. I'm open for API ideas that may help eliminating the last step in future releases. I'm wondering what is your app / use case?

> but it's extremely memory consumptive.

Note that ImageSharp does heavy pooling when managing it's buffers, and with 2.0 this will become much more efficient -- see #1730.

**Reply by sampletext32** [reply]
Actually, I am developing some kind of a video editor and I use ImageSharp for frame processing. So for a something like 30/60 frames per second processing - it becomes really consumprive. The pain occured while I was developing Scale effect. I create a full screen (1920x1080) image and draw another image onto it (this is easily used by Translate effect). So now, I want to scale a whole frame, but I don't want it to exceed the resolution, which is given before. I use backing byte array to communicate with OpenGL, so using full-flex `ImageSharp.Image` and later on convering it to byte array is also very consumptive.
So here came my question.

Maybe there was a `Mutate.DrawImage(Point, Size)` method that I could use. Unfortunately, I will have to do the pixel manipulation manually, because there is definitely an algorithm.

If we Upscale the image, we go from center to bounds, copying pixels.
If we downscale, we go from bounds to center, clearing outer pixels.

**Reply by JimBobSquarePants** [reply]
I don’t see how this is possible. You can’t write where you need to sample from.

**Reply by sampletext32** [reply]
Actually, I have implemented this behavior and it's working.

Here is a video of this algorithm, working in my editor.

https://user-images.githubusercontent.com/44116740/150366799-cd193725-7330-4ac9-9aa2-1c0e0282e40d.mp4

Check my implementation here: [link](https://github.com/SCVE/SCVE/blob/e8d3b401d17d52f273485022da8fdf0e6cdb29b7/SCVE.Editor/Effects/ScaleEffect.cs#L24)

Be aware that this is a singlethreaded solution, so if you can implement a proper multithreaded Resize In-place in ImageSharp, I would really be appreciated.

**Reply by sampletext32** [reply]
Actually, manual pixel copying can eliminate image cloning in translation effect, so that I don't draw an image over existing, but shift the pixels.

**Reply by br3aker** [reply]
It's still possible to get wrong results when sampling from and writing to the same target buffer especially with more advanced sampling methods. 

This can be solved by writing `Mutate` results directly to the target buffer which you use for OpenGL communication. Possible API:
```cs
byte[] buffer = new byte[size]; // byte[] for simplicity
using var img = Image.Load("...");
img.Mutate(
    ctx => ctx
    .ResizeTo(200, 200, target=buffer)
);
```

With `ProcessorName_To_` approach every existing processor which can't work inplace can be used with user provided buffer instead of a internal image copying. But this naming can be misleading though, we are not 'mutating' original image. @antonfirsov

**Reply by JimBobSquarePants** [reply]
That linked implementation looks like nearest neighbor to me and the approach definitely won't work with convolution kernels that use weighted sampling. 

I order to use the same source buffer as a target we'd have to use a separate interim buffer inside the resize processor to copy the source to so it can be sampled. That could theoretically be a window. Obviously it would only work for downsampling. 

If someone wants to play around with that, rock on, but it's not a priority currently for me.
